### PR TITLE
DSEGOG-254 User Sessions Endpoints

### PR DESCRIPTION
**Pull request by MRichards99**
This PR adds 5 new endpoints, to support the user sessions functionality:
- `POST /sessions` - save a session to the database
- `PATCH /sessions/{id}` - update a specific session, given its ID
- `GET /sessions/{id}` - get a specific session by its ID
- `GET /sessions/list` - get a list of the user's sessions
- `DELETE /sessions/{id}` - delete a session by its ID

A user session contains a name, summary, timestamp (generated by the API when a session is inserted into the database), a flag to say whether the session is an auto saved session or if it is manually saved, and the session data itself. The ID is generated when the session is inserted into the database - Object IDs are used as it means they're random and can deal with a user renaming their session (which a concatenation of the username and session name wouldn't be able to do). There's a Pydantic model for this, along with `UserSessionListModel` (as username and session data aren't needed for `/sessions/list`).

I've added two new classes, `UserSession` and `UserSessionList`. `UserSession` is essentially a CRUD class for interacting with sessions to/from the database. For now, `UserSessionList` only needs to retrieve a list of sessions for the current user (so the user can select which session to load) but given a different Pydantic model is used for this, it seemed sensible to separate this into a different class to `UserSession`.

In `src/auth/jwt_handler.py`, I've made a small refactor to put some code into a separate function (`get_payload()`) so that it can be called in the session code. This is so the session code can get the username from the token sent in the request.

I've added a number of unit and integration tests which provide pretty good coverage for the code added; `src/sessions/` has 97% coverage. For the integration tests, I've added another `conftest.py` file, containing two fixtures. This one is stored in the same directory as where the session integration tests are stored and is used to inject and remove test user sessions which are asserted against in the tests. When testing `/sessions/list`, all the user sessions in the database are removed before inserting two test sessions. As this endpoint lists all sessions available to a user, all sessions need to be removed from the database before the request is made to ensure the same starting state is achieved.

I've also updated the Postman collection with the new endpoints, in their own directory called 'User Sessions'.

While this is a 2k+ PR, ~950 lines are caused by `poetry.lock` and there's 600+ lines for tests - the actual code changes aren't that huge (promise!).

### CODE REVIEW

File: operationsgateway_api/src/routes/sessions.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+from operationsgateway_api.src.sessions.user_session_list import UserSessionList
+
+
+log = logging.getLogger()
+router = APIRouter()
+
+
+@router.get(
+    "/sessions/list",
+    summary="Get a list of session names that belong to the given user",
```
[the remark this answers is not in the document]
Comment (MRichards99) [reply operationsgateway_api/src/routes/sessions.py]: Fixed in fea886a0fb309296e7fa88f4202f081716075f4a

File: operationsgateway_api/src/routes/sessions.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+@endpoint_error_handling
+async def get_user_sessions_list(
+    access_token: str = Depends(authorise_token),
+):
+    """
+    Get a list of user sessions that belong to the current user and respond with the
+    sessions' metadata
+    """
+
+    log.info("Getting list of user sessions")
```
[the remark this answers is not in the document]
Comment (MRichards99) [reply operationsgateway_api/src/routes/sessions.py]: Fixed in fea886a0fb309296e7fa88f4202f081716075f4a

File: operationsgateway_api/src/routes/sessions.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        description="Flag to show whether the session has been manually or"
+        " automatically saved",
+    ),
+    access_token: str = Depends(authorise_token),
+):
+    """
+    Save/update a user session (and its metadata) into the database
+    """
+
+    log.info("Saving user session")
```
[the remark this answers is not in the document]
Comment (MRichards99) [reply operationsgateway_api/src/routes/sessions.py]: Fixed in fea886a0fb309296e7fa88f4202f081716075f4a

File: operationsgateway_api/src/sessions/user_session.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            )
+        except InvalidId as exc:
+            raise ModelError("ID provided is not a valid ObjectId") from exc
+
+        log.debug(
+            "Number of sessions matched with id_: %s, number of sessions modified: %s",
+            update_result.matched_count,
+            update_result.modified_count,
+        )
+        if update_result.matched_count != 1:
```
[the remark this answers is not in the document]
Comment (MRichards99) [reply operationsgateway_api/src/sessions/user_session.py:134]: Fixed in fea886a0fb309296e7fa88f4202f081716075f4a

**Comment by MRichards99** [comment]
I've made the requested changes, please could you look at them and re-review the PR?

**Review by kevinphippsstfc — APPROVED** [review]

**Pull request merged**